feat(guidelines): add guidelines for test header usage in rest apis (#65)
otto-de/api-guidelines: api-guidelines/rest/http/headers/rules/may-use-test-header.md

---

> id: R0000XX

Reviewer: ```suggestion
id: R000080
```

Test requests are created in various contexts.
Some of them require special handling on provider or consumer side, such as bypassing specific validations or using modified business logic while processing.
This rule defines the test header that API consumers and providers can use to identify test data.
Downstream teams can then forward this test header in subsequent processing steps, e.g. as a test extension in an asynchronous event.

This has several benefits:

- It allows for testing of certain use cases across multiple teams.
- It enables all teams to identify and ignore hindering test data.
- It ensures data integrity of the overall system.

## `Test` header

The `Test` header is optional but if set, it MUST be a non-empty string in the format `<test-scope>.<test-type>`.
Example: `OrderProcessing.PerformanceTest`.

| Segment  |  Description | Possible values | Constraints  | Example  |
|---|---|---|---|---|
| `test-scope`  | The scope of the executed test. | Possible values include but are not limited to the name of a single context or a group of contexts.  | Must not contain a `.` (dot).  | OrderProcessing  |

> | `test-type`  | The type of the test. | The following common types are defined:<br><br>- `PerformanceTest`: Automatically created test data in large scale.<br> - `SystemTest`: Automatically created test data for test-specific business cases within one team.<br>- `End2EndTest`: Automatically created test data to test specific business cases across multiple teams.<br>- `ManualTest`: Manually created data.<br><br>Additional test types may be used but must explicitly be listed in the API documentation of the API provider. | Must not contain a `.` (dot).  | PerformanceTest  |

Reviewer: Should we update the description for manual test based on our latest learnings? Is it to be used across multiple teams?

Author: Would do that in a separate PR and then change it for the async part aswell.

---
id: R0000XX
---

# MAY use `Test` header

API consumers and providers may use the OTTO-specific test header described within this rule to communicate that a request or response is a result of a test in a **non-live** environment. This enables testing certain use cases together with multiple teams while simultaneously enabling all teams to identify and ignore hindering test data.

> Requests and responses **must not** be flagged with `Test` in the live environment as this could negatively impact the data integrity of the overall system.

Author: This part I took from the async version of this rule because I thought we had the same assumptions about having no test interactions in a live environment. But you are right. If we indeed have tests in a live environment, it would be beneficial to flag those interactions with test headers. Do you have the API performance tests in mind? Does it happen in a live environment as well?
 
I will change the sentence afterwards.

> Requests and responses **must not** be flagged with `Test` in the live environment as this could negatively impact the data integrity of the overall system.

Author: Hey @holgerb, we removed the sentence about test header usage in live environment because, as you mentioned, it has no real benefit of leaving them in live.

> Requests and responses **must not** be flagged with `Test` in the live environment as this could negatively impact the data integrity of the overall system.

Reviewer 2: Solved for me!

Test requests are often created in various contexts. Some of them require special handling on provider or consumer side, such as bypassing specific validations or using modified business logic while processing.

This rule defines the `Test` header that API consumers and providers can provide to identify test data and enable the provider or consumer to forward this test header in their further processing, e.g. passing it as test extension in an asynchronous event.

### `Test` header

> - Type: `String`
> - Description: A string in the format `<test-scope>.<test-type>`. Both `test-type` and `test-scope` must not contain a `.` (dot).
>     - `test-scope`: The scope of the executed test. Possible values include but are not limited to the name of a single context or a group of contexts (e.g., `orderprocessing`).
>     - `test-type`: The type of the test. Currently, the following common types are defined:
>         - `PerformanceTest` Automatically created test data in large scale.
>         - `SystemTest` Automatically created test data to test specific business cases within one team.
>         - `End2EndTest` Automatically created test data to test specific business cases within multiple teams.
>         - `ManualTest` Manually created data.
> 
>       Additional test types may be used. They must be listed in the API documentation of the API provider
> - Examples:
>     - `SampleOrderProcessing.PerformanceTest`
> - Constraints:
>     - OPTIONAL
>     - MUST be a non-empty string in the format `<test-scope>.<test-type>`.

Reviewer: I know that this equals the structure of the async guideline but IMO it's pretty hard to digest. Idea:

```suggestion
The `Test` header is optional but if set, it MUST be a non-empty string in the format `<test-scope>.<test-type>`.
Example: `OrderProcessing.PerformanceTest`.

| Segment  |  Description | Possible values | Constraints  | Example  |
|---|---|---|---|---|
| `test-scope`  | The scope of the executed test. | Possible values include but are not limited to the name of a single context or a group of contexts.  | Must not contain a `.` (dot).  | OrderProcessing  |
| `test-type`  | The type of the test. | The following common types are defined:<br><br>- `PerformanceTest`: Automatically created test data in large scale.<br> - `SystemTest`: Automatically created test data for test-specific business cases within one team.<br>- `End2EndTest`: Automatically created test data to test specific business cases across multiple teams.<br>- `ManualTest`: Manually created data.<br><br>Additional test types may be used but must explicitly be listed in the API documentation of the API provider. | Must not contain a `.` (dot).  | PerformanceTest  |
```

---
id: R000080
---

# MAY use `Test` header

API consumers and providers may use the OTTO-specific `Test` header described within this rule to communicate that a request or response is part of a test.

Test requests are created in various contexts.
Some of them require special handling on provider or consumer side, such as bypassing specific validations or using modified business logic while processing.
This rule defines the test header that API consumers and providers can use to identify test data.

> Downstream teams must then forward this test header in subsequent processing steps, e.g. as a test extension in an asynchronous event.

Reviewer 3: ...in an asynchronous event. > isn't this rule for REST? Shouldn't this read synchronous then?

Author: Yes, but it's possible that API providers sending async events as an direct consequence of a synchronous API call, e.g. place order REST api triggers a "order placed" event. In this case it's also important to include the original test headers in the event.

> Downstream teams must then forward this test header in subsequent processing steps, e.g. as a test extension in an asynchronous event.

Author: Downstream providers and consumers would be possible

This has several benefits:

> - It allows for testing of certain use cases across multiple teams.

Author: teams -> APIs?

> - It enables all teams to identify and ignore hindering test data.

Author: consumers and providers

- It ensures data integrity of the overall system.

## `Test` header

The `Test` header is optional but if set, it MUST be a non-empty string in the format `<test-scope>.<test-type>`.
Example: `OrderProcessing.PerformanceTest`.

| Segment  |  Description | Possible values | Constraints  | Example  |
|---|---|---|---|---|
| `test-scope`  | The scope of the executed test. | Possible values include but are not limited to the name of a single context or a group of contexts.  | Must not contain a `.` (dot).  | OrderProcessing  |

> | `test-type`  | The type of the test. | The following common types are defined:<br><br>- `PerformanceTest`: Automatically created test data in large scale.<br> - `SystemTest`: Automatically created test data for test-specific business cases within one team.<br>- `End2EndTest`: Automatically created test data to test specific business cases across multiple teams.<br>- `ManualTest`: Manually created data.<br><br>Additional test types may be used but must explicitly be listed in the API documentation of the API provider. | Must not contain a `.` (dot).  | PerformanceTest  |

Reviewer 3: "team/teams" > can we generalize this to "provider/consumer" instead of using team/teams?

Is OrderProcessing and PerformanceTest kind of a unique identifier? I'd write it in backticks then.

::: warning Important

> Make sure you get in touch with your consumers and downstream API providers before participating in tests with your APIs in general, as they could impact their system or generate unexpected costs.

Reviewer 3: Hah! This is where you wrote "downstream API providers" ✌🏻, so you can use it above as well.

> Make sure you get in touch with your consumers and downstream API providers before participating in tests with your APIs in general, as they could impact their system or generate unexpected costs.

Reviewer 3: ```suggestion
Make sure you get in touch with your consumers and downstream API providers before participating in tests with your APIs, as these could affect their system or cause unexpected costs.
```

:::

## Example use cases

![Example usages of the test extension](../../../../async/format/test-extension/rules/test-extension-usage-examples.png)

::: references

- [MUST forward `Test` header](./must-forward-test-header.md)

> - [Async event guideline: MAY use test extension](../../../../async/format/test-extension/rules/may-use-test-extension.md)

Reviewer 3: ```suggestion
- [MAY use test extension](../../../../async/format/test-extension/rules/may-use-test-extension.md)
```

> - [Async event guideline: MUST forward `test` context attribute](../../../../async/format/test-extension/rules/must-forward-test-context-attribute.md)

Reviewer 3: ```suggestion
- [MUST forward `test` context attribute](../../../../async/format/test-extension/rules/must-forward-test-context-attribute.md)
```

> - [Async event guideline: MAY ignore events flagged with `test` context attribute](../../../../async/format/test-extension/rules/may-ignore-events-flagged-with-test.md)

Reviewer 3: ```suggestion
- [MAY ignore events flagged with `test` context attribute](../../../../async/format/test-extension/rules/may-ignore-events-flagged-with-test.md)
```
